app/gui2/src/components/ComponentBrowser/input.ts

```diff
@@ -215,44 +227,80 @@ export function useComponentBrowserInput(graphDb: GraphDb = useGraphStore().db)
       (shouldInsertSpace ? ' ' : '') +
       oldCode.substring(newCodeUpToLastChange.oldCodeIndex)
     selection.value = { start: newCursorPos, end: newCursorPos }
+    const context = newCodeUpToLastChange.code
+    if (requiredImport) {
+      const [id] = suggestionDb.nameToId.lookup(requiredImport)
+      if (id) {
+        const requiredEntry = suggestionDb.get(id)
+        if (requiredEntry) {
+          imports.value = imports.value.concat(
+            requiredImports(suggestionDb, requiredEntry).map((info) => ({ info, context })),
+          )
+        }
+      }
+    } else {
+      imports.value = imports.value.concat(
+        requiredImports(suggestionDb, entry).map((info) => ({ info, context })),
+      )
+    }
+  }
+
+  /** List of imports required for applied suggestions.
+   *
+   * If suggestion was manually edited by the user after accepting, it is not included.
+   */
+  function importsToAdd(): RequiredImport[] {
+    const existingImports = graphDb.imports.value
+    const finalImports: RequiredImport[] = []
+    for (const { info, context } of imports.value) {
+      const alreadyImported = existingImports.some((existing) => covers(existing.import, info))
+      const alreadyAdded = finalImports.some((existing) => requiredImportEquals(existing, info))
+      const noLongerNeeded = !code.value.includes(context)
+      if (!alreadyImported && !noLongerNeeded && !alreadyAdded) {
+        finalImports.push(info)
+      }
+    }
+    return finalImports
   }
 
   /** Return all input changes resulting from applying given suggestion.
    *
    * @returns The changes, starting from the rightmost. The `start` and `end` parameters refer
    * to indices of "old" input content.
    */
-  function* inputChangesAfterApplying(
-    entry: SuggestionEntry,
-  ): Generator<{ range: [number, number]; str: string }> {
+  function inputChangesAfterApplying(entry: SuggestionEntry): {
+    changes: { range: ContentRange; str: string }[]
+    requiredImport: QualifiedName | null
+  } {
     const ctx = context.value
     const str = codeToBeInserted(entry)
+    let mainChange = undefined
     switch (ctx.type) {
       case 'insert': {
-        yield { range: [ctx.position, ctx.position], str }
+        mainChange = { range: [ctx.position, ctx.position] as ContentRange, str }
```

Comment:
is this cast actually needed?

Reply:
yeah, afaict it shouldn't be needed. do note that type *assertions* are unsafe, and we should prefer the type-safe `satisfies ContentRange` even if it's needed - but even then, if `satisfies` is necessary, it would be good to add a comment explaining why it is necessary

Reply:
Well, I’m getting this error without the cast:
```
src/components/ComponentBrowser/input.ts:293:24 - error TS2322: Type '{ range: number[]; str: string; }' is not assignable to type '{ range: ContentRange; str: string; }'.
  Types of property 'range' are incompatible.
    Type 'number[]' is not assignable to type 'ContentRange'.
      Target requires 2 element(s) but source may have fewer.

293     return { changes: [mainChange, ...qnChange.changes], requiredImport: qnChange.requiredImport }
                           ~~~~~~~~~~
                        
```
@somebody1234 any ideas?

Reply:
@vitvakatu ah, i didn't notice that it didn't have an explicit type. you'd want to do `let mainChange: { range: ContentRange; str: string } | undefined` (initializer should be optional since `undefined` is the default anyway)

ideally i think the object type should be refactored out into an `interface`, since it's used in multiple methods anyway

Reply:
Fixed.